Various code rearrangements

dcaliste opened the pull request:
@pvuorela, this is a proposition of various code rearragements in mKCal. The main medium-term target would be to make of mKCal a SQLite storage backend only to upstream KCalendarCore objects, by slowly (very slowly) deprecating ExtendedCalendar. But as a short-term consequences these commits should produce an improved decorrelation between the various parts in mKCal and some code cleaning (in my opinion):
- don't impose the calendar of an ExtendedStorage to be an ExtendedStorageObserver. Like that a KCalendarCore::MemoryCalendar or any reimplementation of KCalendarCore::Calendar could become a calendar for our storage backend. Don't imposing this is quite simple in the sense that at the moment, the only consequence of being an observer is to reset itself on storage change. This can actually be imposed from the storage backend itself and not done volontarily in the calendar itself.
- I've taken the liberty to move all SQLite related bits from the public sqlitestorage.h header to the more confidential sqliteformat.h one. This also clean up a bit the cross pointing from SqliteFormat to SqliteStorage.
- I've moved the database location code from ExtendedCalendar to SqliteStorage, because I think it belongs there : the sqlitestorage could propose a default location or take an imposed one like previously. This step is also important for the first point : like that one can create an SQLite backend object based on whatever calendar (and not restricted to ExtendedCalendar one) and still open the existing calendar database.
- last commit, maybe the most controversial : base in ExtendedStorage creation API on KCalendarCore::Calendar instead of ExtendedCalendar. The controversial bit being the `addIncidence()` modifications. We discussed this already before that changing the default calendar when adding an incidence is a bit unfortunate. So let me elaborate.

KCalendarCore::Calendar is not providing a way to add an incidence with a notebook. The upstream way is to add the incidence and then set the notebook. Well, why not… But if we are based on ExtendedCalendar (and for the moment everything is), we cannot add any incidence leaving the notebook empty. The addIncidence() method without notebook, is imposing default notebook. So two choices :
- change ExtendedCalendar behaviour for the addIncidence() not to set any notebook, letting the user to set the notebook. But this is breaking code (and tests) where we suppose that not specifying the notebook will end-up in the default calendar. Particularly when loading from ICS data for instance.
- change the default notebook when adding an incidence. Actually, this is quite painless. We're not touching the "real" default notebook, as exposed by the storage to the outer world, but simply the QString stored in the calendar. If you look at KCalendarCore::Calendar::setDefaultNotebook, it is just copying the string. So, even if a bit ugly in the code, swapping the default notebook is not a dangerous action.

What do you think about these modifications ?

pvuorela commented on src/sqliteformat.cpp:
```diff
@@ -48,9 +47,13 @@ using namespace mKCal;
-        : mStorage(storage), mDatabase(database)
+    Private(SqliteFormat *format, sqlite3 *database, const QTimeZone &timeZone)
+        : mFormat(format), mDatabase(database)
+        , mOriginTime(QDateTime(QDate(1970, 1, 1), QTime(0, 0, 0), Qt::UTC))
```
Not related entirely to these changes but I've been pondering if there's some point in this custom epoch compared to just using QDateTime::toSecsSinceEpoch() / fromSecsSinceEpoch().

dcaliste replied:
I think the point of having a date hard-coded here (and choosing epoch is not a necessity) is because date times in the database are stored as integers corresponding to a shift in seconds to a reference. So one needs to define this reference. What is surprising to me is that this reference is hard-coded instead of being stored in the database. Because at the moment, any change for another reference in the source code will break all existing databases...

But since this reference date time is hard-coded and set to epoch, it could be simpler to actually use the Qt routines converting from epoch directly and drop this definition. I'm going to try...

pvuorela reviewed (commented):
Just some minor comments now but didn't yet check every detail. Overall looks nice.

Interesting that the ExtendedCalendar requirement got away from the storage.

dcaliste commented:
> Interesting that the ExtendedCalendar requirement got away from the storage.

Indeed, I was very suprised, but it's a nice opportunity to make the code have less dependencies. The only real issue is the `addIncidence()` behaviour difference between KCalendarCore::Calendar and ExtendedCalendar.

That being said, thinking about this issue during my commuting time, I'm afraid that the `setDefaultCalendar` policy will work only for ExtendedStorage calendars. Indeed, this will not work with a MemoryCalendar for instance. I'm going to add a test with a simple MemoryCalendar calendar to check that it's indeed independant on the strategy used to setup the notebook.

So I need to find how to make this possible without breaking to much ExtendedCalendar... Let see what I can figure out.

pvuorela commented:
I take that last comment so that you want to check things before we merge. Not considering the last commit the changes look good to me and I could be ok having them already with those couple minor details adjusted.

dcaliste commented:
Indeed, I'm working on a way to get this `addIncidence()` call working for every Calendar implementation, and not only ExtendedCalendar. Here are the requirements:
- one should be able to call `CalStorage::load()` and loaded incidences are assigned to the proper calendar,
- one should be able to call `ICalFormat::fromString()` and loaded incidences are assigned to desired calendar, default or any other specified by the user.

At the moment, the first point is failing since ExtendedCalendar is adding to default calendar on `addIncidence()` while MemoryCalendar is not assigning any calendar on such call. The second point is failing because Upstream is ignoring `notebook` argument in `ICalFormat::fromString()` call, see https://invent.kde.org/frameworks/kcalendarcore/-/merge_requests/98

So I'm currently trying to fix second issue with the upstream PR. To solve the first, I'm going to change the behaviour of ExtendedCalendar::addEvent() in case of one argument call not to save to the default notebook, and keep the call with two argument for this purpose. Like that, ExtendedCalendar behaviour will be aligned with upstream.

> Not considering the last commit the changes look good to me and I could be ok having them already

I agree, I'm going to drop the last commit and create a separate PR for it with the above change in behaviour so it can be discussed decoupled from the rest.

dcaliste commented on src/sqliteformat.cpp:
```diff
@@ -48,9 +47,13 @@ using namespace mKCal;
+        : mFormat(format), mDatabase(database)
+        , mOriginTime(QDateTime(QDate(1970, 1, 1), QTime(0, 0, 0), Qt::UTC))
+        , mTimeZone(timeZone)
+
```
Deleted.

dcaliste commented on src/sqliteformat.cpp:
```diff
@@ -350,9 +355,9 @@ static bool setDateTime(SqliteStorage *storage, sqlite3_stmt *stmt, int &index,
+#define SL3_bind_date_time( format, stmt, index, dt, allDay)          \
     {                                                                  \
-        if (!setDateTime(storage, stmt, index, dt, allDay))            \
+        if (!setDateTime(format, stmt, index, dt, allDay))            \
```
Ah good point, re-aligned.

dcaliste commented:
I've rebased the PR and dropped the commit making ExtendedStorage using KCalendarCore::Calendar instead of ExtendedCalendar. I'm going to open a PR dedicated to this change only.

dcaliste commented:
I've also added a commit dropping `mOriginTime`. Tests seem to pass, so far so good. And it's safer like that in the sense that there is no way to change the reference date time now of the existing databases.
